src/main/java/com/iota/iri/service/transactionpruning/jobs/MilestonePrunerJob.java

```diff
@@ -0,0 +1,262 @@
+     * restored from a state file, because we start cleaning up at the {@link #startingIndex}.
+     *
+     * @param startingIndex milestone index that defines where to start cleaning up
+     */
+    public MilestonePrunerJob(int startingIndex, int targetIndex) {
+        this(startingIndex, startingIndex, targetIndex);
+    }
+
+    /**
+     * Creates a job that cleans all milestones prior (and including) the {@code startingIndex} and that is set to have
+     * progressed already to the given {@code currentIndex}.
+     *
+     * Since cleanup jobs can be consolidated (to reduce the size of the {@link TransactionPruner} state file) and
+     * restored (after IRI restarts), we need to be able provide both parameters even tho the job usually always
+     * "starts" with its {@code currentIndex} being equal to the {@code startingIndex}.
+     *
+     * If the {@link #currentIndex} is bigger than the {@link #targetIndex} we set the state to DONE (necessary when
+     * restoring the job from the state file).
+     *
+     * @param startingIndex milestone index that defines where to start cleaning up
+     * @param currentIndex milestone index that defines the next milestone that should be cleaned up by this job
+     */
+    private MilestonePrunerJob(int startingIndex, int currentIndex, int targetIndex) {
+        setStartingIndex(startingIndex);
+        setCurrentIndex(currentIndex);
+        setTargetIndex(targetIndex);
+
+        if (currentIndex > targetIndex) {
+            setStatus(TransactionPrunerJobStatus.DONE);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * It iterates from the {@link #currentIndex} to the provided {@link #targetIndex} and processes every milestone
+     * one by one. After each step is finished we persist the progress to be able to continue with the current progress
+     * upon IRI restarts.
+     */
+    @Override
+    public void process() throws TransactionPruningException {
+        if (getStatus() != TransactionPrunerJobStatus.DONE) {
+            setStatus(TransactionPrunerJobStatus.RUNNING);
+
+            try {
+                while (!Thread.interrupted() && getStatus() != TransactionPrunerJobStatus.DONE) {
+                    cleanupMilestoneTransactions();
+
+                    setCurrentIndex(getCurrentIndex() + 1);
+
+                    // synchronize this call because the MilestonePrunerJobQueue needs it to check if we can be extended
+                    synchronized (this) {
+                        if (getCurrentIndex() > getTargetIndex()) {
+                            setStatus(TransactionPrunerJobStatus.DONE);
+                        }
+                    }
+
+                    // persist changes if the job was executed by a TransactionPruner (tests might run it standalone)
+                    if (getTransactionPruner() != null) {
+                        getTransactionPruner().saveState();
```

Comment:
its not a bug - the MilestonePrunerJob will run a very long time and process milestones one at a time - everytime it finishes one of these subtasks it persists its progress. I dont want to only save the progress once the whole job is done but whenever it finish cleaning up one of the milestones.

Reply:
hmm:
1. So in this case isn't the outer `saveState` is extraneous. Because once you are done you will still save state in the inner `saveState` and then once more in the outer. I suppose for other types of jobs you have to do it though...
If I understood right one way to overcome the above issue is to add `isSavingStateWhileProcessing` (or something with shorter name) that will return a boolean to `TransactionPrunerJob`. Then you can omit the extra save in the end.

2. There is a slight design problem that `process` implicitly saves state but that is too hard to fix now I think so I will just ignore it. Just FYI

3. I don't like the test "combina" (http://www.ehebrew.org/articles/hebrew-slang.php#.W8yg4XVubdE). In test the job will have to receive a mock or else get an NPE.

Reply:
No the outer saveState is not extraneous because once the job is DONE it will get removed from the Queue and we also need to "save" this. This can not be done from the "inside" because the job doesn't know it is managed by a queue and if and when it will get removed from the Queue.

I am okay with removing this null check and rely on a mock for tests.